### ISSUE TITLE
Numerical issue at triangle edges?

### ISSUE THREAD

**Issue by fynv** [madmann91/bvh#51]
![image](https://user-images.githubusercontent.com/8346412/165441438-bd1e8f98-3ec5-42e7-b92c-6e9b258f3d46.png)

Here I'm using 32bit floats. The test code generates a depth image using SingleRayTraverser. Seems that a few pixels missed intersections at triangle edges. I'm not very sure the is issue is with bvh. Just trying to ask and see if there's any suggestion. Thanks.

**Comment by fynv** [comment]
The issue needs a perfect camera-projection to reproduce. It is gone if I shift the pixels positions by (+0.5, +0.5). So maybe I should consider it as expected behavior..

**Comment by madmann91** [comment]
Is the sphere made out of triangles? If yes, is there a model for it that I can have a look at? I'm asking because I'd like to know if this comes from the geometry not being watertight, or some precision issue in the triangle intersection routine. By the way, are you using the intersection routine from this library, or some other one? And also, are you using the robust intersector or the fast one?

**Comment by fynv** [comment]
I use the following code to create the sphere. Yes, it is made out of triangles. 
```
void GeometryCreator::CreateSphere(Primitive* primitive, float radius, int widthSegments, int heightSegments)
{
	int count_x = widthSegments + 1;
	int count_y = heightSegments + 1;
	int vert_count = count_x * count_y;
	int face_count = widthSegments * heightSegments * 2;

	std::vector<glm::vec4> pos(vert_count);
	std::vector<glm::vec4> norm(vert_count);
	std::vector<glm::vec2> uv(vert_count);
	std::vector<glm::ivec3> faces(face_count);

	for (int j = 0; j < count_y; j++)
	{
		float v = (float)j / (float)heightSegments;
		float phi = (0.5f - v) * (float)PI;
		float cos_phi = cosf(phi);
		float sin_phi = sinf(phi);
		for (int i = 0; i < count_x; i++)
		{
			float u = (float)i / (float)widthSegments;
			float theta = u * 2.0f * (float)PI;
			float cos_theta = cosf(theta);
			float sin_theta = sinf(theta);
			glm::vec3 dir = { cos_phi * sin_theta, sin_phi, cos_phi * cos_theta };

			int idx = i + j * count_x;
			pos[idx] = glm::vec4(dir* radius, 1.0f);
			norm[idx] = glm::vec4(dir, 0.0f);
			uv[idx] = { u,v };
		}
	}

	for (int j = 0; j < heightSegments; j++)
	{
		for (int i = 0; i < widthSegments; i++)
		{
			int a = i + j * count_x;
			int b = (i + 1) + j * count_x;
			int c = i + (j + 1) * count_x;
			int d = (i + 1) + (j + 1) * count_x;

			int idx = (i + j * widthSegments) * 2;
			faces[idx] = { c, b, a };
			faces[idx + 1] = { b, c, d };
		}
	}

	create(primitive, pos, norm, uv, faces);
}
```
The intersector I'm using is: 
bvh::ClosestPrimitiveIntersector<bvh::Bvh<float>, bvh::Triangle<float>>

All of the code can be found at: https://github.com/fynv/Three.V8, specifically:
https://github.com/fynv/Three.V8/blob/main/ThreeEngine/core/BoundingVolumeHierarchy.h
https://github.com/fynv/Three.V8/blob/main/ThreeEngine/core/BoundingVolumeHierarchy.cpp

**Comment by fynv** [comment]
To me, it could be an issue similar to the following case:
a<b<c 
if (x>a && x<b) ...
if (x>b && x<c) ...
when given x == b it misses.

**Comment by madmann91** [comment]
You can first try using `bvh::SingleRayTraverser<Bvh, 64, bvh::RobustNodeIntersector<Bvh>>`. This is a more robust traversal algorithm that should not miss geometry. If geometry is still missed when using this, it could be that the triangle intersector is missing the edges.

**Comment by madmann91** [comment]
I'm not sure if the previous comment helped, but you can also try removing symmetries, perhaps this helps:
```cpp
for (int j = 0; j < heightSegments; j++)
{
    for (int i = 0; i < widthSegments; i++)
    {
        int a = i + j * count_x;
        int b = (i + 1) + j * count_x;
        int c = i + (j + 1) * count_x;
        int d = (i + 1) + (j + 1) * count_x;
        if (j == 0) a = b = 0;
        if (j == heightSegments - 1) c = d = vert_count - 1;
        else if (i == widthSegments - 1)
        {
            b = j * count_x;
            d = (j + 1) * count_x;
        }

        int idx = (i + j * widthSegments) * 2;
        faces[idx] = { c, b, a };
        faces[idx + 1] = { b, c, d };
    }
}
```
Let me know if this does not help.

**Comment by fynv** [comment]
Using "RobustNodeIntersector" doesn't seem to help.
The 2nd trick doesn't work either and it gives me incorrect UV.

In Three.V8 code, now I've enabled the repro by default. Runing Test.V8 or GamePlayer will produce a dmp_depth.raw, (800x800 8bit image).

**Comment by madmann91** [comment]
Alright, I will have a look later today

**Comment by madmann91** [comment]
It's all designed for Windows, so I can't really do anything with the code in your repo. I've successfully ported the ThreeEngine library to CMake, and built it, but the executables all depend on the Win32 API to create windows and I don't really have time to port that to Linux using SDL or something. It would be nice to have an executable that just renders the picture to a file, and terminates immediately.

**Comment by fynv** [comment]
Try this:
```
#include <vector>

#include <bvh/bvh.hpp>
#include <bvh/vector.hpp>
#include <bvh/triangle.hpp>
#include <bvh/sweep_sah_builder.hpp>

#include <bvh/ray.hpp>
#include <bvh/single_ray_traverser.hpp>
#include <bvh/primitive_intersectors.hpp>


const double PI = 3.14159265359;

int main()
{
	int widthSegments = 32;
	int heightSegments = 16;

	int count_x = widthSegments + 1;
	int count_y = heightSegments + 1;
	int vert_count = count_x * count_y;
	int face_count = widthSegments * heightSegments * 2;

	std::vector<float> pos(vert_count*3);
	std::vector<int> faces(face_count*3);

	for (int j = 0; j < count_y; j++)
	{
		float v = (float)j / (float)heightSegments;
		float phi = (0.5f - v) * (float)PI;
		float cos_phi = cosf(phi);
		float sin_phi = sinf(phi);
		for (int i = 0; i < count_x; i++)
		{
			float u = (float)i / (float)widthSegments;
			float theta = u * 2.0f * (float)PI;
			float cos_theta = cosf(theta);
			float sin_theta = sinf(theta);

			int idx = i + j * count_x;
			pos[idx * 3] = cos_phi * sin_theta;
			pos[idx * 3 + 1] = sin_phi;
			pos[idx * 3 + 2] = cos_phi * cos_theta;
		}
	}

	for (int j = 0; j < heightSegments; j++)
	{
		for (int i = 0; i < widthSegments; i++)
		{
			int a = i + j * count_x;
			int b = (i + 1) + j * count_x;
			int c = i + (j + 1) * count_x;
			int d = (i + 1) + (j + 1) * count_x;

			int idx = (i + j * widthSegments) * 2;
			faces[idx * 3] = c;
			faces[idx * 3 + 1] = b;
			faces[idx * 3 + 2] = a;
			faces[idx * 3 + 3] = b;
			faces[idx * 3 + 4] = c;
			faces[idx * 3 + 5] = d;
		}
	}

	typedef bvh::ClosestPrimitiveIntersector<bvh::Bvh<float>, bvh::Triangle<float>> IntersectorType;
	typedef bvh::SingleRayTraverser<bvh::Bvh<float>> TraversorType;

	std::vector<bvh::Triangle<float>> m_triangles;	

	for (int i = 0; i < face_count; i++)
	{
		int i0 = faces[i * 3];
		int i1 = faces[i * 3 + 1];
		int i2 = faces[i * 3 + 2];
		
		float v0_x = pos[i0 * 3];
		float v0_y = pos[i0 * 3 + 1];
		float v0_z = pos[i0 * 3 + 2];

		float v1_x = pos[i1 * 3];
		float v1_y = pos[i1 * 3 + 1];
		float v1_z = pos[i1 * 3 + 2];

		float v2_x = pos[i2 * 3];
		float v2_y = pos[i2 * 3 + 1];
		float v2_z = pos[i2 * 3 + 2];

		m_triangles.emplace_back(bvh::Triangle<float>(
			bvh::Vector3<float>(v0_x, v0_y, v0_z),
			bvh::Vector3<float>(v1_x, v1_y, v1_z),
			bvh::Vector3<float>(v2_x, v2_y, v2_z)
			)
		);
	}

	bvh::BoundingBox<float> m_bounding_box;
	auto [bboxes, centers] = bvh::compute_bounding_boxes_and_centers(m_triangles.data(), m_triangles.size());
	m_bounding_box = bvh::compute_bounding_boxes_union(bboxes.get(), m_triangles.size());

	bvh::Bvh<float> m_bvh;
	bvh::SweepSahBuilder<bvh::Bvh<float>> builder(m_bvh);
	builder.build(m_bounding_box, bboxes.get(), centers.get(), m_triangles.size());

	IntersectorType m_intersector(m_bvh, m_triangles.data());
	TraversorType m_traverser(m_bvh);

	int size = 512;
	float center = 256.0f;
	float step = 0.00546875f;

	std::vector<unsigned char> depth(size * size);

	for (int y = 0; y < size; y++)
	{
		for (int x = 0; x < size; x++)
		{
			float fx = ((float)x - center) * step;
			float fy = ((float)y - center) * step;
			float fz = 2.0f;

			bvh::Ray<float> ray = {
				bvh::Vector3<float>(fx, fy, fz),
				bvh::Vector3<float>(0.0f, 0.0f, -1.0f)				
			};

			unsigned char value = 255;

			auto hit = m_traverser.traverse(ray, m_intersector);
			if (hit.has_value())
			{
				auto intersection = hit->intersection;
				value = (unsigned char)(intersection.t / 3.0f * 255.0f + 0.5f);
			}

			depth[x + y * size] = value;
		}
	}

	FILE* fp = fopen("dmp.raw", "wb");
	fwrite(depth.data(), 1, size* size, fp);
	fclose(fp);


	return 0;
}


```

**Comment by madmann91** [comment]
Thank you! I indeed managed to reproduce the bug and fix it. The bug should be fixed on the master branch, in commit 2fd0db62022993963a7343669275647cb073e19a. Make sure you use `bvh::Triangle<float, true, true>` to get the more robust version of the intersection function, as it is not enabled by default. I also recommend using `bvh::RobustNodeIntersector` instead of the default (`bvh::FastNodeIntersector`) if you want to avoid false negatives (missed intersections). Please re-open if this did not fix the issue for you.
